library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java

```diff
@@ -696,9 +704,83 @@ public boolean isLoading() {
 
   @Override
   public void reevaluateBuffer(long positionUs) {
-    // Do nothing.
+    if (hlsInterruptStateDelegate != null && REQUESTED.equals(hlsInterruptStateDelegate.getCancelState()) && loader.isLoading()) {
+      hlsInterruptStateDelegate.setCancelState(TRIGGERED);
+      loader.interruptLoading();
+      return;
+    }
+    if (loader.isLoading() || isPendingReset()) {
+      return;
+    }
+
+    int currentQueueSize = mediaChunks.size();
+    int preferredQueueSize = chunkSource.getPreferredQueueSize(positionUs, readOnlyMediaChunks);
+    if (currentQueueSize <= preferredQueueSize) {
+      return;
+    }
+
+    int newQueueSize = currentQueueSize;
+    for (int i = preferredQueueSize; i < currentQueueSize; i++) {
+      if (!haveReadFromMediaChunk(i)) {
```

Comment:
I have done most of the cleanup/simplification you've suggested, except for this one:

```
If you want to discard chunks that have already been read from, you need to set the pendingResetPositionUs to the position where playback should continue from (most likely lastSeekPositionUs or the start of the chunk) and clear the list of media chunks. This ensures the renderers get reset correctly. Something like that is already done in upstreamDiscard for the empty queue case, but needs to be extended to cover all cases where we discard chunks that have been read from. If you do this, then you also don't need to change the SampleQueue implementation.
```
This is something I still have to look at, but maybe it's fine for now to only discard chunks that have not been read from.

Do let me know what you think @tonihei

Reply:
@tvarga-dss discarding chunks that have just started playing would be nice (with a 6 second chunk size doing so might avoid up to 6 seconds of poor quality content playing), this of course would need to be balanced against the chance it would cause a stall (very likely).

So, IMHO, start small and only cancel loading chunks that are not currently being rendered.  We saw benefit to this change even without the load cancel, as it is easily possible (in the non-live case at least) for 10s of seconds of poor quality content to be buffered.

Once you have that, I'll re-test to make sure the assert in the sample queue read (https://github.com/google/ExoPlayer/pull/7244#issuecomment-620119891)  has been fixed

Reply:
I agree with Steve's comment - canceling chunks we started reading seems very useful, especially at the beginning of playback. However, it's fine to merge this first and add support as a second step.